Src/WindowsServer/WindowsServer.Shared/AppServicesHeartbeatTelemetryModule.cs

```diff
@@ -0,0 +1,181 @@
+            new KeyValuePair<string, string>("appSrv_wsHost", "WEBSITE_HOSTNAME")
+        };
+
+        // for testing only: override the heartbeat manager
+        internal IHeartbeatPropertyManager HeartbeatManager;
+
+        // to provide a 'singleton' accessor for updating the env vars should they change during runtime
+        private static AppServicesHeartbeatTelemetryModule instance;
+        private bool isInitialized = false;
+        
+        /// <summary>
+        /// Initializes a new instance of the<see cref="AppServicesHeartbeatTelemetryModule" /> class.
+        /// </summary>
+        public AppServicesHeartbeatTelemetryModule() : this(null)
+        {            
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AppServicesHeartbeatTelemetryModule" /> class. This is
+        /// internal, and allows for overriding the Heartbeat Property Manager to test this module with.
+        /// </summary>
+        /// <param name="hbeatPropManager">The heartbeat property manager to use when setting/updating env var values.</param>
+        internal AppServicesHeartbeatTelemetryModule(IHeartbeatPropertyManager hbeatPropManager)
+        {
+            this.HeartbeatManager = hbeatPropManager;
+            AppServicesHeartbeatTelemetryModule.Instance = this;
+        }
+
+        /// <summary>
+        /// Gets a value to provide internal access to the only instance of this class that *should* be available.
+        /// </summary>
+        public static AppServicesHeartbeatTelemetryModule Instance
+        {
+            get => AppServicesHeartbeatTelemetryModule.instance;
+            private set => AppServicesHeartbeatTelemetryModule.instance = value;
+        }
+
+        /// <summary>
+        /// Initialize the default heartbeat provider for Azure App Services. This module
+        /// looks for specific environment variables and sets them into the heartbeat 
+        /// properties for Application Insights, if they exist.
+        /// </summary>
+        /// <param name="configuration">Unused parameter.</param>
+        public void Initialize(TelemetryConfiguration configuration)
+        {
+            this.isInitialized = this.UpdateHeartbeatWithAppServiceEnvVarValues();
+        }
+
+        /// <summary>
+        /// Ensure we've cleaned up our static Instance.
+        /// </summary>
+        public void Dispose()
+        {
+            // ensure our Instance variable is not kept around.
+            AppServicesHeartbeatTelemetryModule.instance = null;
+        }
+
+        /// <summary>
+        /// Signal the AppServicesHeartbeatTelemetryModule to update the values of the 
+        /// Environment variables we use in our heartbeat payload.
+        /// </summary>
+        /// <returns>A value indicating whether or not an update to the environment variables occurred.</returns>
+        public bool UpdateHeartbeatWithAppServiceEnvVarValues()
```

Comment:
Instead of a TelemetryInitializers telling this module to update, can this module update itself (re-read ENV vars) every 'n' minute? And maybe inform the TelemetryInitializer that ENV values changed so that it'll also pick new values.

Reply:
Agreed. I was thinking that a subscriber model would work really nicely actually - that way the heartbeat module and the initializer module could simply 'subscribe' to the environment variable monitor (see PR #870, that's the terminology I use there) and get updated only as necessary. 

For this PR, I wanted to keep it as concise as possible to get the job done. I would certainly be open to creating such a subscriber in our next sprint.

Reply:
After discussions with Dmitry I will create an alternate PR to this one that contains a subscriber model solution. Although due to time constraints we still may opt for this solution.